[3단계] 자동차 경주

RoadtoS7 opened the pull request:
## 구조 설명

- `RandomNumberGenrator`: 0~9사이의 랜덤 문자 생성기 
- `RacingCar`: 자동차에 해당

- `Engine`: RacingCar가 앞으로 갈 수 있는지 결정하는 역할 (프로토콜)
  - `RacingCarEngine`: RandomNumberGenerator를 이용하여 생성한 랜덤 숫자가 4보다 클 경우 이동할 수 있는 Engine,
  Engine 프로토콜을 따릅니다.

- `RacingCarFactory`: RacingCarEngine을 갖는 RacingCar를 생성하는 팩토리

- `NormalRacingRound`:  게임에 참여한 모든 자동차를 전진시키는 역할 
  - `RacingRound`:  Round 마다 참여하는 자동차나, round 진행 규칙이 달라질 수 있을 것이라 가정했습니다.   
  따라서 이 변경에 쉽게 대응하기 위해서 RacingRound 프로토콜을 만들고, NormalRacingRound가 RacingRound프로토콜을 따르도록 구현했습니다.

- `RacingGame`: (시도 횟수)와 RacingCar 객체 리스트를 입력받아서 Round를 생성하고, Round를 시작(start())한다.

- `RacingCar.State`: RacingCar의 id와 location값을 담는 객체
- `RacingRoundResult`: 모든 자동차를 한번씩 전진 시켰을 때 자동차의 상태 값(RacingCar.State)를 담는 객체
- `RacingGameResult`: 모든 Round를 진행한 후 게임 결과를 담는 객체 


## 피드백 요청 사항 
- `RacingCar`를 생성하는 책임을 누구에게 부여해야 할지 고민되었고, 결국 RacingCarFactory에게 이 책임을 부여했어요.  
하지만 `RacingCarFactory` 테스트를 어떻게 해야할지 모르겠습니다.

-  `RacingCar.State`, `RacingRoundResult`, `RacingGameResult` 객체가 꼭 필요한지 고민됩니다.   
이 객체들은 게임 실행 데이터를 담는 역할 밖에 하지 않아서 이렇게 분리하는 것이 오히려 코드를 더 복잡하게 만드는 것 같습니다.

- 테스트 가능한 구조로 짜여졌는지 궁금합니다!

- 제가 누락한 Unit 테스트가 있는지 피드백 부탁드립니다.

- 테스트 커버리지를 어떻게하면 더 높일 수 있을까요? 🤔

피드백 부탁드립니다 🙏

RoadtoS7 committed:
initial commit

RoadtoS7 committed:
feat(RacingCar): Engine에 따라서 움직이는 RacingCar를 구현함

RoadtoS7 committed:
feat(RacingCar): 자동차 경주 게임을 시도하는 각 횟수를 나타내는 round 구현

RoadtoS7 committed:
feat(RacingCarGame): 자동차 경주 게임을 나타내는 RacingCarGame 구현

- RacingCar는 자동차 경주에 참여할 RacingCar와 round횟수를 입력받아서 생성된다
- Round 구성방식은 자동차 경주 게임마다 달라지므로 자동차 경주 게임에서 Round 객체를 생성하는 책임을 갖도록 했다

RoadtoS7 committed:
refactor(RacingCarTest): 항상 움직이는 자동차 객체를 생성할 때 MovingEngine 객체를 사용하도록 수정

기존에는 RacingCarTest안에서 AlwaysMovingEngine을 생성하여 사용했었다
하지만 다른 UnitTest에서도 반복되어서 사용함에 따라 MoginEngine이라는 Fake 객체를 만들어서 사용한다

RoadtoS7 committed:
refactor(RacingCarTest): 자동차가 움직이지 않을때 테스트케이스 추가

RoadtoS7 committed:
refactor(RacingRound): RacingRound프로토콜을 만들어 RacingGame과 RacingRound사이에 DIP를 준수하도록 수정:

RoadtoS7 committed:
style: 줄바꿈

RoadtoS7 committed:
feat(InputView): 사용자의 입력을 받는 InputView 구현

RoadtoS7 committed:
feat(GameResultFormatter): 게임 결과를 출력 양식에 맞게 변환하는 GameResultFormatter 구현

RoadtoS7 committed:
feat: 게임 결과를 출력하는 OutputView 구현

RoadtoS7 committed:
feat: 사용자의 입력이 유효한지 검사하는 InputChecker 구현

자동차 대수와 시도할 횟수는 양의 정수여야 함을 검사한다.

RoadtoS7 committed:
feat: 입력값이 잘못되었을 때 오류 메시지 출력하는 기능을 OutputView에 추가

RoadtoS7 committed:
feat(main): 프로그램 통합 로직 구현

RoadtoS7 committed:
style: InputView, OutputView, GameResultFomatter를 View로 폴더링함

RoadtoS7 committed:
refator: View 구성요소의 targetMembership에 RacingCarTest 타겟을 추가함

RoadtoS7 committed:
refactor: InputChecker의 targetMembership에 RacingCarTest타겟을 추가함

RoadtoS7 committed:
refactor: RacingCar를 생성하는 역할을 하는 RacingCarFactory 구현

RoadtoS7 committed:
refactor: GameResultFormatter에서 함수가 하나의 일만 하도록 수정

RoadtoS7 committed:
refactor: 각 자동차의 기록을 저장하는 역할을 RacingCar.Stat에게 부여함e

RoadtoS7 committed:
fix: 시도할 횟수만큼 자동차가 생성되는 이슈 해결

RoadtoS7 committed:
refactor: RacingRoundResult의 생성자 수정

RacingGameTest를 테스트하기 쉽게 만들기 위해서 RacingRoudnResult가 RacingCar.State를 받아서 생성되도록 수정했다

RoadtoS7 committed:
refactor: RacingCar에서 필요없는 프로퍼티 제거

RoadtoS7 committed:
feature: InputChecker에 대한 테스트 코드 구현

RoadtoS7 committed:
style: 오타 수정

RoadtoS7 committed:
refactor: RacingRound unit test 수정

RacingRound 수행 결과에서 RacingCar의 id 값도 올바른지 테스트하도록 수정

uuu1101 commented on RacingCar/RacingCar.swift:
```diff
@@ -0,0 +1,47 @@
+    }
+    
+    struct State: Equatable {
+        let id: Int
```
id를 두신 이유가 궁금합니다.

RoadtoS7 replied:
피드백 감사합니다! 👏

![스크린샷 2022-05-02 오후 10 51 31](https://user-images.githubusercontent.com/57262833/166245227-5c82ff83-248a-44f4-8595-56de39ce47a5.jpg)

출력때 어떤 RacingCar에 대한 기록인지 알 수 있도록 하기 위해서 State에서 RacingCar의 id도 보관하고자 했습니다! 
근데 구현을 확인해보니 사용하지 않은 데이터네요😅
없애고 다시 구현해보겠습니다.

RoadtoS7 replied:
현재 RacingCarFactory처럼 난수 생성 로직이나 생성된 난수가 4보다 큰 지 체크하는 로직을 따로 클래스로 빼놓는게 좋다고 생각합니다.
이 부분에 대한 테스트가 꼭 필요한것인지? 그것이 아니라면 커버리지를 채우기 위한 테스트 코드를 작성 할 것인지? 에 대해서도 고민해볼 필요가 있다고 생각합니다.

저의 코드에서 RacingCarFactory는 RacingCar 인스턴스를 생성하는 역할만 하는데, "특정 타입의 인스턴스 생성"은 검증이 필요없을 만큼 명확한 행위이기 때문에 테스트가 꼭 필요하지는 않은 것 같습니다..!
태형님께서는 어떻게 생각하시는지 궁금합니다!!

추가적으로 개선하면 좋은 구조나 클래스가 있을까요?
그리고 코드 컨벤션이 잘 지켜졌고, 네이밍이 적절하지 않은 부분이 있는지도 확인부탁드립니다!👀

uuu1101 reviewed (commented):
안녕하세요 ! Step3 고생하셨습니다 👏👏👏

- Car의 Engine을 표현하신게 인상 깊었어요 😊
---
- RacingCar를 생성하는 책임을 누구에게 부여해야 할지 고민되었고, 결국`RacingCarFactory`에게 이 책임을 부여했어요.
하지만 `RacingCarFactory` 테스트를 어떻게 해야할지 모르겠습니다.
> 현재 RacingCarFactory처럼 난수 생성 로직이나 생성된 난수가 4보다 큰 지 체크하는 로직을 따로 클래스로 빼놓는게 좋다고 생각합니다.
이 부분에 대한 `테스트가 꼭 필요한것인지?` 그것이 아니라면 `커버리지를 채우기 위한 테스트 코드를 작성 할 것인지?` 에 대해서도 고민해볼 필요가 있다고 생각합니다.

- `RacingCar.State`, `RacingRoundResult`, `RacingGameResult` 객체가 꼭 필요한지 고민됩니다.
이 객체들은 게임 실행 데이터를 담는 역할 밖에 하지 않아서 이렇게 분리하는 것이 오히려 코드를 더 복잡하게 만드는 것 같습니다.

> 의견을 드리자면, 단순히 데이터를 담는 역할 밖에 하지않는다면 분리하지 않고 해당 데이터가 필요한 객체에 둬도 괜찮을 것 같습니다. RacingCar의 경우 id와 location 처럼 따로 State로 분리하지 않고 RacingCar가 가져도 될 것 같다는 생각입니다.

- 테스트 가능한 구조로 짜여졌는지 궁금합니다!
- 제가 누락한 Unit 테스트가 있는지 피드백 부탁드립니다.
- 테스트 커버리지를 어떻게하면 더 높일 수 있을까요? 🤔

> 세 가지 모두 비슷한 질문인것 같아, 같이 답변 드립니다.
앞서 말씀드린것 처럼 create 메서드, 즉 난수 생성 및 조건 체크 하는 부분에 대해 테스트가 꼭 필요한 것인지? 처럼 확인해볼 필요가 있다고 생각합니다. 
자동차 대수가 0이거나 시도할 횟수를 0일 경우등 조건이 달라지는 경우에도 테스트를 수행하도록 해볼 수 있을 것 같습니다.

다른 의견이 있으시다면 코멘트 남겨주세요 😊

RoadtoS7 commented:
> 현재 RacingCarFactory처럼 난수 생성 로직이나 생성된 난수가 4보다 큰 지 체크하는 로직을 따로 클래스로 빼놓는게 좋다고 생각합니다.
이 부분에 대한 테스트가 꼭 필요한것인지? 그것이 아니라면 커버리지를 채우기 위한 테스트 코드를 작성 할 것인지? 에 대해서도 고민해볼 필요가 있다고 생각합니다.

저의 코드에서 RacingCarFactory는 RacingCar 인스턴스를 생성하는 역할만 하는데,  "특정 타입의 인스턴스 생성"은 검증이 필요없을 만큼 명확한 행위이기 때문에 테스트가 꼭 필요하지는 않은 것 같습니다..! 
태형님께서는 어떻게 생각하시는지 궁금합니다!!

uuu1101 commented on RacingCar/InputChecker.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct InputChecker {
+    static func isValid(carCount: Int?) -> Bool {
```
```suggestion
    static func isValidCarCount(_ count: Int?) -> Bool {
```
아래 isValid라는 메서드 명이 중복되어서(물론 파라미터로 구분이 가능할테지만), 이런식으로 명확하게 나눠주는건 어떨까요? 👀

uuu1101 commented on RacingCar/InputChecker.swift:
```diff
@@ -0,0 +1,18 @@
+        return carCount != nil && carCount! > 0
+    }
+    
+    static func isValid(roundCount: Int?) -> Bool {
```
```suggestion
    static func isValidRoundCount(_ count: Int?) -> Bool {
```
위 코멘트처럼 의견드려봅니다 🙇🏻‍♂️

uuu1101 commented on RacingCar/main.swift:
```diff
@@ -1,12 +1,27 @@
-let result = try Calculator().calculate(operators: operations, operands: operands)
-OutputView.show(output: result)
+let carCount = InputView.inputCarCount()
+if !InputChecker.isValid(carCount: carCount) {
```
```suggestion
if InputChecker.isValid(carCount: carCount) == false {
```
저는 개인적으로 위처럼 명시하는 것을 선호합니다 😊

uuu1101 reviewed (changes requested):
고생하셨습니다 ! 👍🏻

- 저의 코드에서 RacingCarFactory는 RacingCar 인스턴스를 생성하는 역할만 하는데, "특정 타입의 인스턴스 생성"은 검증이 필요없을 만큼 명확한 행위이기 때문에 테스트가 꼭 필요하지는 않은 것 같습니다..!
태형님께서는 어떻게 생각하시는지 궁금합니다!!
> 저도 동일한 생각을 가지고 있습니다 😊 로직을 체크하는 것도 따로 클래스로 뺴두신 부분도 마찬가지구요!

네이밍이나 컨벤션에 관해서 피드백을 달라고 하셨는데, 개인의 선호가 반영되는 부분도 있는거 같아, 의견만 드려봅니다 ! 👀
크리티컬하게 문제되는 내용은 아니지만, Foundation을 사용하지 않는데 import하신 부분도 있네요.😊

수정이 필요하다고 생각되지 않으시면 코멘트로 태그하셔서 merge 요청해주시면 됩니다! 🙇🏻‍♂️

RoadtoS7 committed:
RacingCar, RacingRound, RacingCar에서 각각 게임 결과를 보관하도록 함

RoadtoS7 committed:
코드 리뷰 반영

1. InputChecker에서 메서드들이 파라미터 이름으로 구분되던 것을 메서드 명으로 구분되도록 수정했다.
2. main함수에서 InputChecker의 결과 값이 false인지 검사하는 if문을 == false를 사용하도록 수정했다.
3. 실행결과를 담는 클래스를 따로 두는 것이 아니라 RacingCar, RacingRound에서 각각 갖도록 했다.

RoadtoS7 committed:
RacingGame이 1개 이상의 RacingCar와 roundCount를 갖도록 수정

RoadtoS7 committed:
RacingRound는 1개 이상의 RacingCar만 가질 수 있도록 수정

RoadtoS7 commented:
안녕하세요 태형님~
네이밍 컨벤션과 관련된 피드백을 적용해보았는데 태형님이 주신 피드백을 따르니 코드가 더 읽기 쉬워졌습니다 감사합니다~
1차 피드백때 주신 대로 RacingCar와 RacingRound에서 게임 결과를 가지고 있도록 수정했습니다.
 
**그런데 이렇게 수정하는 과정에서 RacingCar를 struct에서 class로 수정해야 했으며, 참조 타입에서 값복사가 일어나도록 하기 위해 copy() 메서드를 사용하게 되었습니다. 이로인해 코드가 복잡해진 것 같아서,,, 이것을 개선하고 싶은데 어떻게 해야할지 잘 모르겠습니다. copy() 메서드 사용하는 로직을 어떻게 개선하면 좋을지 피드백 부탁드립니다! 🙏**  

(struct를 class로 수정하게 된 이유는 RacingCound에서 RacingCar의 move(mutating method)를 호출해주어야 하는데, RacingCar가 struct일 경우, 호출할 수 없었기 때문입니다.)

그리고 **RacingCar에서 Equatable 프로토콜의 `==` 메서드 내용을 구현할 때, location값과 engine 타입을 비교하여 같으면 true를 반환하고자 했으나 engine의 타입을 비교하는 방법을 잘 모르겠습니다.** 결국 engine을 제외하고 location만 같으면 같은 인스턴스로 판정하기에는 로직상 말이 안되었기에 `id` 프로퍼티를 두고 이것을 비교하는 방향으로 구현했습니다. 이 부분도 어떻게 개선하면 좋을지 피드백 부탁드립니다.

마지막으로 RacingGame의 결과로 `[RacingRound]` 를 반환하도록 했는데 이 부분이 어색한 것 같아서 RacingGame을 반환하는 것이 더 좋을지 고민되었습니다..  둘중 어느 방법이 더 좋다고 생각하시는지 태형님의 생각이 궁금합니다.

RoadtoS7 commented on RacingCar/RacingCar.swift:
```diff
@@ -0,0 +1,46 @@
+    
+    let id: Int
+    let engine: Engine
+    private var _location = 0
+    var location: Int { return _location }
```
location 프로퍼티를 RacingCar에서만 변경할 수 있고, 외부에서는 read-only 접근할 수 있도록 하기 위해 이렇게 구현했는데 혹시 더 좋은 방법이 있을지 궁금합니다 🙏

uuu1101 replied:
현재 구조에서는 RacingCar의 move를 통해서 location을 변경하기 때문에 private으로 설정하여 외부에서 read-only 접근할 수 있도록 하신건 아주 좋은것 같아요 :)
차에 엔진을 두어서 실생활과 유사하게 객체를 가져가신 것도 아주 좋았습니다.

제가 요구사항을 보고 생각해본 방향은 Round에서도 차를 옮겨 줄수 있다고도 한번 생각해보았어요 😊
뭔가 이런 의견을 드릴때마다 더 혼란스럽게만 해드리는건 아닌지 조심스럽네요 👀

uuu1101 commented:
고생많으셨습니다. 다음 스텝을 진행하면서 같이 더 고민해보도록 해요 😊

그런데 이렇게 수정하는 과정에서 RacingCar를 struct에서 class로 수정해야 했으며, 참조 타입에서 값복사가 일어나도록 하기 위해 copy() 메서드를 사용하게 되었습니다. 이로인해 코드가 복잡해진 것 같아서,,, 이것을 개선하고 싶은데 어떻게 해야할지 잘 모르겠습니다. copy() 메서드 사용하는 로직을 어떻게 개선하면 좋을지 피드백 부탁드립니다! 🙏

> 여기서 개선이라는게 코드가 복잡해지지 않는걸 원하시는 걸까요?.. 
struct는 사용할 수 없고, 값 복사를 위해 copy()를 사용해야하는데, 코드가 복잡해지는것이 문제라면 struct를 사용 하지 못하게 된 구조적인 문제를 찾아보아야 할것 같다는 생각이 듭니다. 
RacingCar와 RacingRound에서 게임 결과를 가지고 있도록 수정하는 과정에서 class로 꼭 변경이 되어야하고 copy()를 사용하는것이 복잡하고 로직을 따라가는데 어려움을 주게 된다면 이전 구조로 가져가는게 더 나은 방향이 될수도 있다고 생각되네요.
RacingCar를 struct로 가져가면서 car의 위치를 변경할수 있는 방법은 없을까요?  
 
 RacingCar에서 Equatable 프로토콜의 == 메서드 내용을 구현할 때, location값과 engine 타입을 비교하여 같으면 true를 반환하고자 했으나 engine의 타입을 비교하는 방법을 잘 모르겠습니다. 
 > 이렇게 비교 하시려고 하는 이유를 알고 싶습니다 🧐
 
 결국 engine을 제외하고 location만 같으면 같은 인스턴스로 판정하기에는 로직상 말이 안되었기에 id 프로퍼티를 두고 이것을 비교하는 방향으로 구현했습니다. 
  >  id 프로퍼티를 두고 비교하게 되면 발생하는 문제가 있을까요?
